### PR TITLE
gh-122697: Fix free-threading memory leaks at shutdown

### PR DESCRIPTION
**Pull request by colesbury**
We were not properly accounting for interpreter memory leaks at shutdown and had two sources of leaks:

 * Objects that use deferred reference counting and were reachable via static types outlive the final GC. We now disable deferred reference counting on all objects if we are calling the GC due to interpreter shutdown.

 * `_PyMem_FreeDelayed` did not properly check for interpreter shutdown so we had some memory blocks that were enqueued to be freed, but never actually freed.

* Issue: gh-122697

### CODE REVIEW

**Comment by colesbury** [comment]
!buildbot nogil refleaks

**Comment by Eclips4** [comment]
The refleak buildbots haven't started yet, but I suspect there are a lot of leaks due to #122704, so the results will not be reliable until fix (#122705) is merged.

**Comment by colesbury** [comment]
!buildbot nogil refleaks

File: Objects/obmalloc.c [hunk 1]
```diff
@@ -1109,9 +1109,9 @@ free_delayed(uintptr_t ptr)
 #ifndef Py_GIL_DISABLED
     free_work_item(ptr);
 #else
-    if (_PyRuntime.stoptheworld.world_stopped) {
-        // Free immediately if the world is stopped, including during
-        // interpreter shutdown.
+    if (Py_IsFinalizing() || _PyRuntime.stoptheworld.world_stopped) {
+        // Free immediately during interpreter shutdown or if the world is
```
Comment (kumaraditya303) [inline Objects/obmalloc.c:1116]: runtime shutdown would be better instead of interpreter shutdown?

Comment (colesbury) [reply Objects/obmalloc.c:1116]: Thanks, this should be interpreter shutdown. That's why the refleak buildbots had failures.

**Comment by colesbury** [comment]
!buildbot nogil refleaks

File: Misc/NEWS.d/next/Core_and_Builtins/2024-08-05-19-28-12.gh-issue-122697.17MvYl.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Fixed memory leaks at interpreter shutdown in the free-threaded build, and
+also reporting of leaked memory blocks via ``-X showrefcount``.
```
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-08-05-19-28-12.gh-issue-122697.17MvYl.rst]:
```suggestion
also reporting of leaked memory blocks via :option:`-X showrefcount <-X>`.
```

**Comment by encukou** [comment]
+1: I don't know my way around this code, but I don't see anything wrong with the PR.

**Review by DinoV — APPROVED** [review]
LGTM!

**Pull request merged**